File: 236. Lowest Common Ancestor of a Binary Tree/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def lowestCommonAncestor(self, root, p, q):
        """
        :type root: TreeNode
        :type p: TreeNode
        :type q: TreeNode
        :rtype: TreeNode
        """
        lca, _ = self.LCA(root, p, q)
        return lca
        
    def LCA(self, root, p, q):
        if not root:
            return None, False
        
        # look left and right
        left, lstatus = self.LCA(root.left, p, q)
        right, rstatus = self.LCA(root.right, p, q)
        
        # if already found, pass along
        if lstatus:
            return left, lstatus
        if rstatus:
            return right, rstatus
        
        # check if p and q found in this subtree
        cand = set([left, right, root])
        if p in cand and q in cand:
            return root, True
        elif (p in cand and not q in cand):
            return p, False
        elif not p in cand and q in cand:
            return q, False
        else:
            return None, False

File: 236. Lowest Common Ancestor of a Binary Tree/solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        Result r = search(root, p, q);
        return r.node;
    }
    
    public Result search(TreeNode root, TreeNode p, TreeNode q) {
        Result s = new Result();
        if (root == null) {
            return s;
        }
        
        Result l = search(root.left, p, q);
        if (l.node != null) {
            return l;
        }
        Result r = search(root.right, p, q);
        if (r.node != null) {
            return r;
        }
        
        if (root.val == p.val) {
            s.foundp = true;
        } else if (root.val == q.val) {
            s.foundq = true;
        }
        
        s.foundp = s.foundp || l.foundp || r.foundp;
        s.foundq = s.foundq || l.foundq || r.foundq;
        if (s.foundp && s.foundq) {
            s.node = root;
        }
        
        return s;
    }
    
    class Result {
        boolean foundp;
        boolean foundq;
        TreeNode node;
        
        public Result() {
            foundp = false;
            foundq = false;
            node = null;
        }
    }
        

}
